pkg/imager/cache/cache.go

```diff
@@ -0,0 +1,189 @@
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+// Package cache provides facilities for generating a cache tarball from images.
+package cache
+
+import (
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"github.com/google/go-containerregistry/pkg/authn"
+	"github.com/google/go-containerregistry/pkg/authn/github"
+	"github.com/google/go-containerregistry/pkg/crane"
+	"github.com/google/go-containerregistry/pkg/name"
+	v1 "github.com/google/go-containerregistry/pkg/v1"
+	"github.com/google/go-containerregistry/pkg/v1/empty"
+	"github.com/google/go-containerregistry/pkg/v1/mutate"
+	"github.com/google/go-containerregistry/pkg/v1/remote"
+	"github.com/google/go-containerregistry/pkg/v1/tarball"
+	"github.com/google/go-containerregistry/pkg/v1/types"
+
+	"github.com/siderolabs/talos/pkg/imager/filemap"
+)
+
+const (
+	blobsDir     = "blob"
+	manifestsDir = "manifests"
+)
+
+// Generate generates a cache tarball from the given images.
+//
+//nolint:gocyclo,cyclop
+func Generate(images []string, platform string, insecure bool, dest string) error {
+	v1Platform, err := v1.ParsePlatform(platform)
+	if err != nil {
+		return fmt.Errorf("parsing platform: %w", err)
+	}
+
+	tmpDir, err := os.MkdirTemp("", "talos-image-cache-gen")
+	if err != nil {
+		return fmt.Errorf("creating temporary directory: %w", err)
+	}
+
+	defer os.RemoveAll(tmpDir) //nolint:errcheck
+
+	if err := os.MkdirAll(filepath.Join(tmpDir, blobsDir), 0o755); err != nil {
+		return err
+	}
+
+	nameOptions := []name.Option{
+		name.StrictValidation,
+	}
+
+	craneOpts := []crane.Option{
+		crane.WithAuthFromKeychain(
+			authn.NewMultiKeychain(
+				authn.DefaultKeychain,
+				github.Keychain,
+			),
+		),
+	}
+
+	remoteOpts := []remote.Option{
+		remote.WithAuthFromKeychain(authn.NewMultiKeychain(
+			authn.DefaultKeychain,
+			github.Keychain,
+		)),
+		remote.WithPlatform(*v1Platform),
+	}
+
+	if insecure {
+		craneOpts = append(craneOpts, crane.Insecure)
+		nameOptions = append(nameOptions, name.Insecure)
+	}
+
+	for _, src := range images {
+		ref, err := name.ParseReference(src, nameOptions...)
+		if err != nil {
+			return fmt.Errorf("parsing reference %q: %w", src, err)
+		}
+
+		if err := os.MkdirAll(filepath.Join(tmpDir, manifestsDir, ref.Context().RegistryStr(), ref.Context().RepositoryStr(), "reference"), 0o755); err != nil {
+			return err
+		}
+
+		if err := os.MkdirAll(filepath.Join(tmpDir, manifestsDir, ref.Context().RegistryStr(), ref.Context().RepositoryStr(), "digest"), 0o755); err != nil {
+			return err
+		}
+
+		manifest, err := crane.Manifest(
+			src,
+			craneOpts...,
+		)
+		if err != nil {
+			return fmt.Errorf("fetching manifest %q: %w", src, err)
+		}
+
+		rmt, err := remote.Get(
+			ref,
+			remoteOpts...,
+		)
+		if err != nil {
+			return fmt.Errorf("fetching image %q: %w", src, err)
+		}
+
+		if !strings.HasPrefix(ref.Identifier(), "sha256:") {
```

Comment:
I wonder if we should handle the reference as 

`some/image:v1.3.5@sha256:abcdef` (pinned ref), and write both tag & sha versions of the digest

Reply:
so we enforce it has a digest as input always?

Reply:
no, not enforce, but handle that correctly.